components/style/properties/helpers/animated_properties.mako.rs

```diff
@@ -757,6 +766,695 @@ impl Interpolate for LengthOrNone {
+            skew.2 = dot(row[1], row[2]);
+            row[2] = combine(row[2], row[1], 1.0, -skew.2);
+
+            // Next, get Z scale and normalize 3rd row.
+            let row2len = (row[2][0] * row[2][0] + row[2][1] * row[2][1] + row[2][2] * row[2][2]).sqrt();
+            scale.2 = row2len;
+            row[2] = [row[2][0] / row1len, row[2][1] / row1len, row[2][2] / row1len];
+            skew.1 /= scale.2;
+            skew.2 /= scale.2;
+
+            // At this point, the matrix (in rows) is orthonormal.
+            // Check for a coordinate system flip.  If the determinant
+            // is -1, then negate the matrix and the scaling factors.
+            let pdum3 = cross(row[1], row[2]);
+            if dot(row[0], pdum3) < 0.0 {
+                % for i in range(3):
+                    scale.${i} *= -1.0;
+                    row[${i}][0] *= -1.0;
+                    row[${i}][1] *= -1.0;
+                    row[${i}][2] *= -1.0;
+                % endfor
+            }
+
+            // Now, get the rotations out
+            let mut quaternion = Quaternion (
+                0.5 * ((1.0 + row[0][0] - row[1][1] - row[2][2]).max(0.0)).sqrt(),
+                0.5 * ((1.0 - row[0][0] + row[1][1] - row[2][2]).max(0.0)).sqrt(),
+                0.5 * ((1.0 - row[0][0] - row[1][1] + row[2][2]).max(0.0)).sqrt(),
+                0.5 * ((1.0 + row[0][0] + row[1][1] + row[2][2]).max(0.0)).sqrt()
+            );
+
+            if row[2][1] > row[1][2] {
+                quaternion.0 = -quaternion.0
+            }
+            if row[0][2] > row[2][0] {
+                quaternion.1 = -quaternion.1
+            }
+            if row[1][0] > row[0][1] {
+                quaternion.2 = -quaternion.2
+            }
+
+            MatrixDecomposed3D {
+                translate: translate,
+                scale: scale,
+                skew: skew,
+                perspective: perspective,
+                quaternion: quaternion
+            }
+        }
+    }
+
+    // Combine 2 point.
+    fn combine(a: [f32; 3], b: [f32; 3], ascl: f32, bscl: f32) -> [f32; 3] {
+        [
+            (ascl * a[0]) + (bscl * b[0]),
+            (ascl * a[1]) + (bscl * b[1]),
+            (ascl * a[2]) + (bscl * b[2])
+        ]
+    }
+
+    // Dot product.
+    fn dot(a: [f32; 3], b: [f32; 3]) -> f32 {
+        a[0] * b[0] + a[1] * b[1] + a[2] * b[2]
+    }
+
+    // Cross product.
+    fn cross(row1: [f32; 3], row2: [f32; 3]) -> [f32; 3] {
+        [
+            row1[1] * row2[2] - row1[2] * row2[1],
+            row1[2] * row2[0] - row1[0] * row2[2],
+            row1[0] * row2[1] - row1[1] * row2[0]
+        ]
+    }
+
+    impl Interpolate for MatrixDecomposed3D {
+        /// https://drafts.csswg.org/css-transforms/#interpolation-of-decomposed-3d-matrix-values
+        fn interpolate(&self, other: &Self, time: f64) -> Result<Self, ()> {
+            let dot = self.quaternion.0 * other.quaternion.0 +
```

Comment:
This doesn't interpolate all the other fields of `MatrixDecomposed3D`. Spec says "Each component of the decomposed values translation, scale, skew and perspective of the source matrix get linearly interpolated with each corresponding component of the destination matrix.", and only explicitly provides steps for quaternions, but the rest needs to be handled too.

Create a copy of `self` called `interpolated`, which has all fields but the last interpolated, and then fiddle with that last field.